README.md
# Simple Clock
<img alt="Logo" src="graphics/icon.png" width="120" />

This clock app has multiple functions related to timing. It can be used as a clock widget or as a alarm clock. It is made to help you regulate your daily life and sleep better. You can also use the stopwatch in this app to count your time when you are running for healthy lifestyle or for any other purpose. This app can also be placed on your home screen for easy navigation.

As a clock widget you can enable displaying times from other time zones, or use the simple, but customizable and resizable clock widget. The text color of the clock widget can be customized, as well as the color and the alpha of the background. You can also change the shape of clock widget according to your choice and show it on the home screen.

The alarm contains all the expected features as day selecting, vibration toggling, ringtone selecting, snooze or adding a custom label. Waking up will be a pleasure. It supports as many alarms as you want, so there won't be any more excuses for not waking up and sleep better :) Gradual volume increasing is supported too, enabled by default. A customizable Snooze button is available too, just in case you really had a good reason for using it. The alarm clock provided by this app as simple as it can get. You simply have to add how many times you want and turn them on. During this, you can also take help from a guide built in this alarm clock app to help you navigate through this app to sleep better. You can sleep better so this app can wake you up on the set time without disturbing your lifestyle. This alarm can be placed on the home screen to make it easy for you to access the alarm while you can work on other things on your device. The main goal of keeping the alarm in this clock widget is to help you schedule your time more effectively. 

With the stopwatch you can easily measure a longer period of time, or individual laps. You can sort the laps in a few different ways. It contains optional vibrations on button presses too, just to let you know that the button was pressed in case you cannot look at the device for some reason, or you are in a hurry. This stopwatch can help you getting in shape if you are doing yoga or having a run in the park. You can put the stopwatch on home screen so that you can easily access it and alter it according to your needs without opening the menu and finding it.

You can easily setup a timer to be notified of some events. You can both change its ringtone, or toggle vibrations. You will never burn that pizza again. The timer countdown can be paused too, not just stopped.

Additional features include for example preventing the device from falling asleep while the app is in foreground or toggling between 12 or 24 hour time format. Last but not least you can decide if the week should start on Sunday, or Monday.

It comes with material design and dark theme by default, provides great user experience for easy usage. The lack of internet access gives you more privacy, security and stability than other apps. The dark theme in this clock widget can help you setting your alarm clock at night without blinding your eyes with sharp color of your mobile alarm.

It comes with material design and dark theme by default, provides great user experience for easy usage. The lack of internet access gives you more privacy, security and stability than other apps.

Contains no ads or unnecessary permissions. It is fully opensource, provides customizable colors.

Check out the full suite of Simple Tools here:  
https://www.simplemobiletools.com

Facebook:  
https://www.facebook.com/simplemobiletools

Reddit:  
https://www.reddit.com/r/SimpleMobileTools

Telegram:  
https://t.me/SimpleMobileTools

<a href='https://play.google.com/store/apps/details?id=com.simplemobiletools.clock'><img src='https://simplemobiletools.com/images/button-google-play.svg' alt='Get it on Google Play' height='45' /></a>
<a href='https://f-droid.org/packages/com.simplemobiletools.clock'><img src='https://simplemobiletools.com/images/button-f-droid.png' alt='Get it on F-Droid' height='45' /></a>

<div style="display:flex;">
<img alt="App image" src="fastlane/metadata/android/en-US/images/phoneScreenshots/1_en-US.jpeg" width="30%">
<img alt="App image" src="fastlane/metadata/android/en-US/images/phoneScreenshots/2_en-US.jpeg" width="30%">
<img alt="App image" src="fastlane/metadata/android/en-US/images/phoneScreenshots/3_en-US.jpeg" width="30%">
</div>



CHANGELOG.md
Changelog
==========

Version 5.11.2 *(2023-09-23)*
----------------------------

 * Fixed some alarm related glitches
 * Added some UI, translation and stability improvements

Version 5.11.1 *(2023-09-18)*
----------------------------

 * Fixed some alarm related glitches
 * Added some UI, translation and stability improvements

Version 5.11.0 *(2023-09-13)*
----------------------------

 * Fixed some alarm related glitches
 * Added some UI, translation and stability improvements

Version 5.10.3 *(2023-07-09)*
----------------------------

 * Fixed next alarm displaying on the widget

Version 5.10.2 *(2023-07-09)*
----------------------------

 * Allow dismissing upcoming alarms early
 * Fixed wrong alarm being shown as next in some cases
 * Fixed duplicate alarm sound on some devices
 * Added some UI, translation and stability improvements

Version 5.10.1 *(2023-03-25)*
----------------------------

 * Added some UI, translation and stability improvements

Version 5.10.0 *(2022-12-26)*
----------------------------

 * Use Material You theme by default on Android 12+
 * Added some UI, translation and stability improvements

Version 5.9.5 *(2022-12-12)*
----------------------------

 * Added a Start Stopwatch shortcut
 * Added some translation, stability and UX improvements

Version 5.9.4 *(2022-10-14)*
----------------------------

 * Changed the default digital widget colors to White text on transparent background
 * Added some stability and translation improvements

Version 5.9.3 *(2022-10-09)*
----------------------------

 * Adding a Get Simple Phone button into the About section
 * Added some translation, stability and UX improvements

Version 5.9.2 *(2022-09-02)*
----------------------------

 * Adding some crashfixes on lower OS versions

Version 5.9.1 *(2022-09-01)*
----------------------------

 * Adding some crashfixes on lower OS versions

Version 5.9.0 *(2022-08-31)*
----------------------------

 * Lowered the minimal required OS version back to Android 6
 * Added Monochrome icon and Language picker support on Android 13+
 * Added some translation, stability and UX improvements

Version 5.8.1 *(2022-07-03)*
----------------------------

 * Make alarms more reliable
 * Adding many translation, stability and UX improvements

Version 5.8.0 *(2022-06-24)*
----------------------------

 * Allow sorting alarms by both day and time
 * Moved tabs at the bottom
 * Added some stability, translation and UX improvements

Version 5.7.1 *(2022-05-19)*
----------------------------

 * Make the date display on the widget more reliable
 * Added some stability and translation improvements

Version 5.7.0 *(2022-04-25)*
----------------------------

 * Increased the minimal required Android version to 10
 * Added Material You theme support on Android 12+
 * Added an analogue clock widget
 * Redesigned the widget config screens
 * Made the widget and alarms more reliable
 * Added some stability, translation and UI improvements

Version 5.6.1 *(2021-11-19)*
----------------------------

 * Redesigned some screens, added an Auto theme
 * Some stability, translation and UI improvements

Version 5.6.0 *(2021-09-15)*
----------------------------

 * Allow creating multiple timers
 * Added some UI and translation improvements

Version 5.5.1 *(2021-04-19)*
----------------------------

 * Allow sorting alarms by creation order or alarm time
 * Added some stability and translation improvements

Version 5.5.0 *(2020-12-29)*
----------------------------

 * Added some stability and translation improvements
 * Require a Simple Thank You purchase for color customization

Version 5.4.0 *(2020-11-05)*
----------------------------

 * Allow creating alarms without any day set
 * Fix last alarm or clock item not being visible at the bottom
 * Added some other UX, stability and translation improvements

Version 5.3.0 *(2020-08-19)*
----------------------------

 * Fixed alarm not vibrating in some cases (by rien)
 * Some minor translation and UI improvements

Version 5.2.2 *(2020-05-28)*
----------------------------

 * Use rounded corners at the widget
 * Properly reschedule new alarm when dismissed
 * Added some other UI and translation improvements

Version 5.2.1 *(2020-03-20)*
----------------------------

 * Removing some extra unwanted permissions

Version 5.2.0 *(2020-03-18)*
----------------------------

 * Make sure the timer doesn't stop if the app gets closed
 * Try making the widget more reliable
 * Adding a label field  to the timer
 * Remember last used Alarm related values, use them at new ones
 * Couple other UI and translation improvements

Version 5.1.0 *(2020-02-14)*
----------------------------

 * Fixed a glitch with all alarms turning on/off wrongly
 * Properly display the alarm snooze screen on Android 8+ (by coxtor)
 * Some stability and translation improvements

Version 5.0.4 *(2019-10-31)*
----------------------------

 * Adding some stability and translation improvements

Version 5.0.3 *(2019-08-07)*
----------------------------

 * Try using vectors instead of some pngs to reduce app size
 * Some smaller improvements here and there

Version 5.0.2 *(2019-07-06)*
----------------------------

 * Some translation improvements

Version 5.0.1 *(2019-04-08)*
----------------------------

 * Some translation improvements

Version 5.0.0 *(2019-01-24)*
----------------------------

 * Increased the minimal required Android OS version to 5
 * Fixed a few alarm related glitches

Version 4.3.0 *(2018-09-19)*
----------------------------

 * Rewrote the widget to scale text size dynamically on Android 8+
 * Couple stability and translation improvements

Version 4.2.0 *(2018-08-10)*
----------------------------

 * Show a timer notification if the app gets paused while it runs
 * Couple other UX improvements

Version 4.1.0 *(2018-07-08)*
----------------------------

 * Adding a toggle for gradual volume increasing
 * Rework the Alarm screen to use swipe to left/right for snooze/dismiss

app/src/main/kotlin/com/simplemobiletools/clock/helpers/Constants.kt
package com.simplemobiletools.clock.helpers

import com.simplemobiletools.clock.extensions.isBitSet
import com.simplemobiletools.clock.models.MyTimeZone
import com.simplemobiletools.commons.helpers.*
import java.util.Calendar
import java.util.Date
import java.util.TimeZone
import kotlin.math.pow

// shared preferences
const val SELECTED_TIME_ZONES = "selected_time_zones"
const val EDITED_TIME_ZONE_TITLES = "edited_time_zone_titles"
const val TIMER_SECONDS = "timer_seconds"
const val TIMER_VIBRATE = "timer_vibrate"
const val TIMER_SOUND_URI = "timer_sound_uri"
const val TIMER_SOUND_TITLE = "timer_sound_title"
const val TIMER_CHANNEL_ID = "timer_channel_id"
const val TIMER_LABEL = "timer_label"
const val TOGGLE_STOPWATCH = "toggle_stopwatch"
const val TIMER_MAX_REMINDER_SECS = "timer_max_reminder_secs"
const val ALARM_MAX_REMINDER_SECS = "alarm_max_reminder_secs"
const val ALARM_LAST_CONFIG = "alarm_last_config"
const val TIMER_LAST_CONFIG = "timer_last_config"
const val INCREASE_VOLUME_GRADUALLY = "increase_volume_gradually"
const val ALARMS_SORT_BY = "alarms_sort_by"
const val STOPWATCH_LAPS_SORT_BY = "stopwatch_laps_sort_by"
const val WAS_INITIAL_WIDGET_SET_UP = "was_initial_widget_set_up"

const val TABS_COUNT = 4
const val EDITED_TIME_ZONE_SEPARATOR = ":"
const val ALARM_ID = "alarm_id"
const val NOTIFICATION_ID = "notification_id"
const val DEFAULT_ALARM_MINUTES = 480
const val DEFAULT_MAX_ALARM_REMINDER_SECS = 300
const val DEFAULT_MAX_TIMER_REMINDER_SECS = 60
const val SIMPLE_PHONE = "Simple_Phone"
const val ALARM_NOTIFICATION_CHANNEL_ID = "Alarm_Channel"
const val EARLY_ALARM_DISMISSAL_CHANNEL_ID = "Early Alarm Dismissal"

const val OPEN_STOPWATCH_TAB_INTENT_ID = 9993
const val PICK_AUDIO_FILE_INTENT_ID = 9994
const val REMINDER_ACTIVITY_INTENT_ID = 9995
const val OPEN_ALARMS_TAB_INTENT_ID = 9996
const val OPEN_APP_INTENT_ID = 9997
const val ALARM_NOTIF_ID = 9998
const val TIMER_RUNNING_NOTIF_ID = 10000
const val STOPWATCH_RUNNING_NOTIF_ID = 10001
const val EARLY_ALARM_DISMISSAL_INTENT_ID = 10002
const val EARLY_ALARM_NOTIF_ID = 10003

const val OPEN_TAB = "open_tab"
const val TAB_CLOCK = 0
const val TAB_ALARM = 1
const val TAB_STOPWATCH = 2
const val TAB_TIMER = 3
const val TIMER_ID = "timer_id"
const val INVALID_TIMER_ID = -1

// stopwatch sorting
const val SORT_BY_LAP = 1
const val SORT_BY_LAP_TIME = 2
const val SORT_BY_TOTAL_TIME = 4

// alarm sorting
const val SORT_BY_CREATION_ORDER = 0
const val SORT_BY_ALARM_TIME = 1
const val SORT_BY_DATE_AND_TIME = 2

const val TODAY_BIT = -1
const val TOMORROW_BIT = -2

// stopwatch shortcut
const val STOPWATCH_SHORTCUT_ID = "stopwatch_shortcut_id"
const val STOPWATCH_TOGGLE_ACTION = "com.simplemobiletools.clock.TOGGLE_STOPWATCH"

val DAY_BIT_MAP = mapOf(
    Calendar.SUNDAY to SUNDAY_BIT,
    Calendar.MONDAY to MONDAY_BIT,
    Calendar.TUESDAY to TUESDAY_BIT,
    Calendar.WEDNESDAY to WEDNESDAY_BIT,
    Calendar.THURSDAY to THURSDAY_BIT,
    Calendar.FRIDAY to FRIDAY_BIT,
    Calendar.SATURDAY to SATURDAY_BIT,
)

fun getDefaultTimeZoneTitle(id: Int) = getAllTimeZones().firstOrNull { it.id == id }?.title ?: ""

fun getPassedSeconds(): Int {
    val calendar = Calendar.getInstance()
    val isDaylightSavingActive = TimeZone.getDefault().inDaylightTime(Date())
    var offset = calendar.timeZone.rawOffset
    if (isDaylightSavingActive) {
        offset += TimeZone.getDefault().dstSavings
    }
    return ((calendar.timeInMillis + offset) / 1000).toInt()
}

fun formatTime(showSeconds: Boolean, use24HourFormat: Boolean, hours: Int, minutes: Int, seconds: Int): String {
    val hoursFormat = if (use24HourFormat) "%02d" else "%01d"
    var format = "$hoursFormat:%02d"

    return if (showSeconds) {
        format += ":%02d"
        String.format(format, hours, minutes, seconds)
    } else {
        String.format(format, hours, minutes)
    }
}

fun getTomorrowBit(): Int {
    val calendar = Calendar.getInstance()
    calendar.add(Calendar.DAY_OF_WEEK, 1)
    val dayOfWeek = (calendar.get(Calendar.DAY_OF_WEEK) + 5) % 7
    return 2.0.pow(dayOfWeek).toInt()
}

fun getTodayBit(): Int {
    val calendar = Calendar.getInstance()
    val dayOfWeek = (calendar.get(Calendar.DAY_OF_WEEK) + 5) % 7
    return 2.0.pow(dayOfWeek).toInt()
}

fun getBitForCalendarDay(day: Int): Int {
    return DAY_BIT_MAP[day] ?: 0
}

fun getCurrentDayMinutes(): Int {
    val calendar = Calendar.getInstance()
    return calendar.get(Calendar.HOUR_OF_DAY) * 60 + calendar.get(Calendar.MINUTE)
}

fun getAllTimeZones() = arrayListOf(
    MyTimeZone(1, "GMT-11:00 Midway", "Pacific/Midway"),
    MyTimeZone(2, "GMT-10:00 Honolulu", "Pacific/Honolulu"),
    MyTimeZone(3, "GMT-09:00 Anchorage", "America/Anchorage"),
    MyTimeZone(4, "GMT-08:00 Los Angeles", "America/Los_Angeles"),
    MyTimeZone(5, "GMT-08:00 Tijuana", "America/Tijuana"),
    MyTimeZone(6, "GMT-07:00 Phoenix", "America/Phoenix"),
    MyTimeZone(7, "GMT-07:00 Chihuahua", "America/Chihuahua"),
    MyTimeZone(8, "GMT-07:00 Denver", "America/Denver"),
    MyTimeZone(9, "GMT-06:00 Costa Rica", "America/Costa_Rica"),
    MyTimeZone(10, "GMT-06:00 Chicago", "America/Chicago"),
    MyTimeZone(11, "GMT-06:00 Mexico City", "America/Mexico_City"),
    MyTimeZone(12, "GMT-06:00 Regina", "America/Regina"),
    MyTimeZone(13, "GMT-05:00 Bogota", "America/Bogota"),
    MyTimeZone(14, "GMT-05:00 New York", "America/New_York"),
    MyTimeZone(15, "GMT-04:00 Caracas", "America/Caracas"),
    MyTimeZone(16, "GMT-04:00 Barbados", "America/Barbados"),
    MyTimeZone(17, "GMT-04:00 Halifax", "America/Halifax"),
    MyTimeZone(18, "GMT-04:00 Manaus", "America/Manaus"),
    MyTimeZone(19, "GMT-03:30 St. John's", "America/St_Johns"),
    MyTimeZone(20, "GMT-03:00 Santiago", "America/Santiago"),
    MyTimeZone(21, "GMT-03:00 Recife", "America/Recife"),
    MyTimeZone(22, "GMT-03:00 Sao Paulo", "America/Sao_Paulo"),
    MyTimeZone(23, "GMT-03:00 Buenos Aires", "America/Buenos_Aires"),
    MyTimeZone(24, "GMT-03:00 Nuuk", "America/Godthab"),
    MyTimeZone(25, "GMT-03:00 Montevideo", "America/Montevideo"),
    MyTimeZone(26, "GMT-02:00 South Georgia", "Atlantic/South_Georgia"),
    MyTimeZone(27, "GMT-01:00 Azores", "Atlantic/Azores"),
    MyTimeZone(28, "GMT-01:00 Cape Verde", "Atlantic/Cape_Verde"),
    MyTimeZone(29, "GMT+00:00 Casablanca", "Africa/Casablanca"),
    MyTimeZone(30, "GMT+00:00 Greenwich Mean Time", "Etc/Greenwich"),
    MyTimeZone(31, "GMT+00:00 London", "Europe/London"),
    MyTimeZone(32, "GMT+01:00 Amsterdam", "Europe/Amsterdam"),
    MyTimeZone(33, "GMT+01:00 Belgrade", "Europe/Belgrade"),
    MyTimeZone(34, "GMT+01:00 Brussels", "Europe/Brussels"),
    MyTimeZone(35, "GMT+01:00 Madrid", "Europe/Madrid"),
    MyTimeZone(36, "GMT+01:00 Sarajevo", "Europe/Sarajevo"),
    MyTimeZone(37, "GMT+01:00 Brazzaville", "Africa/Brazzaville"),
    MyTimeZone(38, "GMT+02:00 Windhoek", "Africa/Windhoek"),
    MyTimeZone(39, "GMT+02:00 Amman", "Asia/Amman"),
    MyTimeZone(40, "GMT+02:00 Athens", "Europe/Athens"),
    MyTimeZone(41, "GMT+02:00 Istanbul", "Europe/Istanbul"),
    MyTimeZone(42, "GMT+02:00 Beirut", "Asia/Beirut"),
    MyTimeZone(43, "GMT+02:00 Cairo", "Africa/Cairo"),
    MyTimeZone(44, "GMT+02:00 Helsinki", "Europe/Helsinki"),
    MyTimeZone(45, "GMT+02:00 Jerusalem", "Asia/Jerusalem"),
    MyTimeZone(46, "GMT+02:00 Harare", "Africa/Harare"),
    MyTimeZone(47, "GMT+03:00 Minsk", "Europe/Minsk"),
    MyTimeZone(48, "GMT+03:00 Baghdad", "Asia/Baghdad"),
    MyTimeZone(49, "GMT+03:00 Moscow", "Europe/Moscow"),
    MyTimeZone(50, "GMT+03:00 Kuwait", "Asia/Kuwait"),
    MyTimeZone(51, "GMT+03:00 Nairobi", "Africa/Nairobi"),
    MyTimeZone(52, "GMT+03:30 Tehran", "Asia/Tehran"),
    MyTimeZone(53, "GMT+04:00 Baku", "Asia/Baku"),
    MyTimeZone(54, "GMT+04:00 Tbilisi", "Asia/Tbilisi"),
    MyTimeZone(55, "GMT+04:00 Yerevan", "Asia/Yerevan"),
    MyTimeZone(56, "GMT+04:00 Dubai", "Asia/Dubai"),
    MyTimeZone(57, "GMT+04:30 Kabul", "Asia/Kabul"),
    MyTimeZone(58, "GMT+05:00 Karachi", "Asia/Karachi"),
    MyTimeZone(59, "GMT+05:00 Oral", "Asia/Oral"),
    MyTimeZone(60, "GMT+05:00 Yekaterinburg", "Asia/Yekaterinburg"),
    MyTimeZone(61, "GMT+05:30 Kolkata", "Asia/Kolkata"),
    MyTimeZone(62, "GMT+05:30 Colombo", "Asia/Colombo"),
    MyTimeZone(63, "GMT+05:45 Kathmandu", "Asia/Kathmandu"),
    MyTimeZone(64, "GMT+06:00 Almaty", "Asia/Almaty"),
    MyTimeZone(65, "GMT+06:30 Rangoon", "Asia/Rangoon"),
    MyTimeZone(66, "GMT+07:00 Krasnoyarsk", "Asia/Krasnoyarsk"),
    MyTimeZone(67, "GMT+07:00 Bangkok", "Asia/Bangkok"),
    MyTimeZone(68, "GMT+07:00 Jakarta", "Asia/Jakarta"),
    MyTimeZone(69, "GMT+08:00 Shanghai", "Asia/Shanghai"),
    MyTimeZone(70, "GMT+08:00 Hong Kong", "Asia/Hong_Kong"),
    MyTimeZone(71, "GMT+08:00 Irkutsk", "Asia/Irkutsk"),
    MyTimeZone(72, "GMT+08:00 Kuala Lumpur", "Asia/Kuala_Lumpur"),
    MyTimeZone(73, "GMT+08:00 Perth", "Australia/Perth"),
    MyTimeZone(74, "GMT+08:00 Taipei", "Asia/Taipei"),
    MyTimeZone(75, "GMT+09:00 Seoul", "Asia/Seoul"),
    MyTimeZone(76, "GMT+09:00 Tokyo", "Asia/Tokyo"),
    MyTimeZone(77, "GMT+09:00 Yakutsk", "Asia/Yakutsk"),
    MyTimeZone(78, "GMT+09:30 Darwin", "Australia/Darwin"),
    MyTimeZone(79, "GMT+10:00 Brisbane", "Australia/Brisbane"),
    MyTimeZone(80, "GMT+10:00 Vladivostok", "Asia/Vladivostok"),
    MyTimeZone(81, "GMT+10:00 Guam", "Pacific/Guam"),
    MyTimeZone(82, "GMT+10:00 Magadan", "Asia/Magadan"),
    MyTimeZone(83, "GMT+10:30 Adelaide", "Australia/Adelaide"),
    MyTimeZone(84, "GMT+11:00 Hobart", "Australia/Hobart"),
    MyTimeZone(85, "GMT+11:00 Sydney", "Australia/Sydney"),
    MyTimeZone(86, "GMT+11:00 Noumea", "Pacific/Noumea"),
    MyTimeZone(87, "GMT+12:00 Majuro", "Pacific/Majuro"),
    MyTimeZone(88, "GMT+12:00 Fiji", "Pacific/Fiji"),
    MyTimeZone(89, "GMT+13:00 Auckland", "Pacific/Auckland"),
    MyTimeZone(90, "GMT+13:00 Tongatapu", "Pacific/Tongatapu")
)

fun getTimeUntilNextAlarm(alarmTimeInMinutes: Int, days: Int): Int? {
    val calendar = Calendar.getInstance()
    calendar.firstDayOfWeek = Calendar.MONDAY
    val currentTimeInMinutes = calendar.get(Calendar.HOUR_OF_DAY) * 60 + calendar.get(Calendar.MINUTE)
    val currentDayOfWeek = calendar.get(Calendar.DAY_OF_WEEK) - Calendar.MONDAY

    var minTimeDifferenceInMinutes = Int.MAX_VALUE

    for (i in 0..6) {
        val alarmDayOfWeek = (currentDayOfWeek + i) % 7
        if (isAlarmEnabledForDay(alarmDayOfWeek, days)) {
            val timeDifferenceInMinutes = getTimeDifferenceInMinutes(currentTimeInMinutes, alarmTimeInMinutes, i)
            if (timeDifferenceInMinutes < minTimeDifferenceInMinutes) {
                minTimeDifferenceInMinutes = timeDifferenceInMinutes
            }
        }
    }

    return if (minTimeDifferenceInMinutes != Int.MAX_VALUE) {
        minTimeDifferenceInMinutes
    } else {
        null
    }
}

fun isAlarmEnabledForDay(day: Int, alarmDays: Int) = alarmDays.isBitSet(day)

fun getTimeDifferenceInMinutes(currentTimeInMinutes: Int, alarmTimeInMinutes: Int, daysUntilAlarm: Int): Int {
    val minutesInADay = 24 * 60
    val minutesUntilAlarm = daysUntilAlarm * minutesInADay + alarmTimeInMinutes
    return if (minutesUntilAlarm > currentTimeInMinutes) {
        minutesUntilAlarm - currentTimeInMinutes
    } else {
        minutesInADay - (currentTimeInMinutes - minutesUntilAlarm)
    }
}
